### ISSUE TITLE
Connection not closed properly

### ISSUE THREAD

**Issue by orangesoup** [actix/actix-net#351]
I've been using 1.x version of actix-web for months, had to restart my app every now and then (sometimes after minutes, sometimes after days) since there are a lot of `ESTABLISHED` connections left there hanging, eventually causing `too many open files` error (I've increased the limit drastically). I'm using my server with keep-alive disabled, the rest of the settings are the defaults. I have since tried to upgrade to 2.0.0 to see if it solves the problem, but it's the same thing.

The service itself gets around 500-1000 requests per second in production currently.

**Comment by fakeshadow** [comment]
https://github.com/actix/actix-net/blob/b2e96409525739c2a6675dfddeefb604977bac28/actix-server/src/accept.rs#L470

os error like 23 /24 would trigger a deregister of socket from mio. There is no drain of socket backlog after they are re-registered.

**Comment by fakeshadow** [comment]
I made a possible fix with this branch https://github.com/actix/actix-net/tree/fix/os_error_23.

If you have this issue please help and give it a test.

I would give some example code on how to test actix-server directly and you need these deps in your `Cargo.toml`
```
actix-rt = { version = "2.2", default-features = false }
actix-server = { git = "https://github.com/actix/actix-net.git", branch = "fix/os_error_23" }
actix-service = "2"
env_logger = "0.8"
tokio = { version = "1.5", features = ["io-util"] }
```

```rust
use std::io;

use actix_rt::net::TcpStream;
use actix_server::Server;
use actix_service::fn_service;
use tokio::io::AsyncWriteExt;

const BUF: &[u8] =
    b"HTTP/1.1 200 OK\r\n\
    content-length: 12\r\n\
    connection: close\r\n\
    date: Thu, 01 Jan 1970 12:34:56 UTC\r\n\
    \r\n\
    Hello World!";

fn main() -> io::Result<()> {
    std::env::set_var("RUST_LOG", "actix=trace, info");
    env_logger::init();

    let name = "short_life";
    let addr = "127.0.0.1:8080";

    actix_rt::System::new().block_on(async {
        Server::build()
            // You want to try lower the backlog if the issue presist and observe if
            // it reduce the hanging connnection count.
            .backlog(2048)
            .bind(name, addr, || fn_service(response))?
            .run()
            .await
    })
}

async fn response(mut stream: TcpStream) -> io::Result<()> {
    stream.write(BUF).await?;
    stream.flush().await?;
    stream.shutdown().await
}
```

**Comment by Ten0** [comment]
> I made a possible fix with this branch https://github.com/actix/actix-net/tree/fix/os_error_23.
> 
> If you have this issue please help and give it a test.

I can still create `Os { code: 24, kind: Other, message: "Too many open files" }` by running `wrk -t7 -c60000 -d180 http://localhost:8000/healthcheck` on my server with:
```toml
[patch.crates-io]
actix-server = { git = "https://github.com/actix/actix-net.git", branch = "fix/os_error_23" }
```
Notably, pretty quickly a large number of https://github.com/actix/actix-web/blob/c17662fe39523ce4230234004c572618932be2c7/src/response/builder.rs#L425 start flooding my console as well.

**Comment by skerkour** [comment]
How can we be sure it comes from Actix and not from a bad configuration of the server (https://www.cyberciti.biz/faq/linux-increase-the-maximum-number-of-open-files/)?

**Comment by fakeshadow** [comment]
@Ten0 
I think you are confused with what this issue is about.

An OS error is an error by you OS. There is no way to remove the error except you fix your OS setting. In your case when you run out of file descriptor you always get "Too many open files".

The issue and the fix is about not producing hanged TCP connections. And you should test this instead of reading just the log.

**Comment by Ten0** [comment]
> @Ten0
> I think you are confused with what this issue is about.
> 
> An OS error is an error thrown by you OS. There is no way to remove the error except you fix your OS setting or ignore it which could result in a busy loop in polling thread. In your case when you run out of file descriptor you always get "Too many open files".

Our issue is that our servers, when starting to receive a few hundred of queries a minute, die in two minutes after startup by hitting "too many open files" errors. It does not look like the OS configuration could cause a so quick crash (its ulimit is at ~1024, so it should make Actix able to handle that rate of connections), which is why I suspect it's a connection leak.

Even if that was not the case, shouldn't Actix should avoid taking every available open file from the OS, making the program unable to do anything else?

I'm having a hard time reproducing the exact same symptoms we get in production (when we receive a valid request, we make other requests that eventually fail because of too-many-open-files), but I suspect this is still the corresponding issue, especially considering [this test](https://github.com/actix/actix-net/issues/351#issuecomment-731865120), which is why I'm attempting to reproduce it using the tools suggested above.

(Note: [our production system still runs on actix-web 2](https://github.com/actix/actix-net/issues/351#issuecomment-734455348))

**Comment by fakeshadow** [comment]
Like I said an OS error is just an OS error. It's returned by `mio` and which I believe would straight up source to syscalls to like using `select` or `epoll`(Or other implementation provided by your OS).

`actix-server` does nothing special when receiving them.  When an OS error that considered non fatal is produced
it delay retry on polling again for 500ms so you don end up polling the same error in a busy loop that would take 100% of your CPU resource. Other than this it does nothing to the error and it can't if it want to.

This issue is about TCP connection leaks and if you have other issue please open a new one. 

And like I said if you can re-produce the leak please give a report on how the "fix" branch behaivor. How many connections have leaked? Does it have differ from the release build you use? Did you try to change the backlog like I pointed and see if it have effect on the amount of leak of connections?

**Comment by jokermario** [comment]
@Ten0  @fakeshadow
while investigating the cause of the problem "too many open files" I stumbled on this [issue](https://github.com/hyperium/hyper/issues/1422) on the hyperium repository, obviously they have had the same issue. According to this [comment](https://github.com/hyperium/hyper/issues/1422#issuecomment-462475590), this is a upstream bug in some versions of libc. He proceeded to put a link to the [issue](https://github.com/rust-lang/rust/issues/47955). This is a [comment](https://github.com/rust-lang/rust/issues/47955#issuecomment-460663710) in that issue. Tell me what you think.

**Comment by finnbear** [comment]
I'm experiencing the same issue, with this set of versions:
```toml
actix = "0.12"
actix-codec = "0.4.0"
actix-web = {version="4.0.0-beta.9", features=["rustls"]}
actix-web-actors = "4.0.0-beta.7"
```

My `actix-web` server runs on stock Debian 11 and hosts both REST and websocket endpoints using `rustls`-based TLS. It leaks about 100 `ESTABLISHED` TCP connections every day (out of an estimated ~10,000-100,000 connections), and has already crashed once because of this. When I run `netstat -nat`, I see that the vast majority of connections have a 0 send-q and recv-q. A small minority have a few 100-1000 for the send-q.

I doubt this is relevant but I have my `keep_alive` set to `KeepAlive::Timeout(30)`.

What can I do to help troubleshoot?
- For various reasons, including this, I must restart my server regularly. What code should I add/change to fix/debug this issue?
- I would try the potential fix at https://github.com/actix/actix-net/tree/fix/os_error_23, but that link is now broken.
- FWIW, my [server](https://github.com/SoftbearStudios/mk48/blob/1d161bb134f94e51c10b701666b62a7ca4e11bf5/server/src/main.rs#L187-L309) and [terraform configuration](https://github.com/SoftbearStudios/mk48/tree/main/terraform) are open source, so anyone could theoretically reproduce the issue by deploying identical infrastructure if the simple example proposed above does not yield a reproduction (it would take a few hours, though)

**Comment by finnbear** [comment]
Here is a reproducible procedure and (nearly) minimal example: https://github.com/finnbear/actix-tcp-leak

They key to reproducing the bug is to establish a connection from another host, send a request, and then terminate the underlying internet connection (i.e. WiFi in the case of a laptop). I observed that such a connection will be leaked for hours (maybe forever), but far longer than the keepalive period I set.

I will now work to further minimize and containerize the example.
